### PR TITLE
feat(db, webserver): Support user avatars

### PR DESCRIPTION
**Pull request by boxbeam**
Fixes TAB-541

Usage: Frontend must encode avatars being uploaded as base64

cc @liangfung

### CODE REVIEW

File: ee/tabby-db/migrations/0019_user-avatar.up.sql
Comment (wsxiaoys) [inline ee/tabby-db/migrations/0019_user-avatar.up.sql:1]: use blob

File: ee/tabby-webserver/src/schema/mod.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:422]:
```suggestion
    async fn upload_user_avatar_base64(ctx: &Context, id: ID, avatar_base64: String) -> Result<bool> {
```

File: ee/tabby-webserver/src/schema/auth.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/auth.rs:247]:
```suggestion
```

Add http api interface (accessible by any login user) `/avatar/{user_id}` to return avatar image

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

**Comment by liangfung** [comment]
cc @wwayne

**Comment by boxbeam** [comment]
Still need to test locally before merging

File: ee/tabby-db/src/users.rs
Comment (wsxiaoys) [inline ee/tabby-db/src/users.rs:23]: Blob itself might not be valid utf-8 string, storing it as `String` will likely generate panic in runtime.

File: ee/tabby-webserver/src/schema/mod.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:428]: this doesn't compile. Please always test your PR locally before marking it ready for review.

File: ee/tabby-webserver/src/hub/mod.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/hub/mod.rs:54]: this shouldn't be part of `hub` crate.

I suggest move the function to `src/handler.rs` and orchestrate the middleware there.

File: ee/tabby-webserver/src/handler.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/handler.rs:80]:
```suggestion
            .nest("/avatar/:id", ...)
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: ee/tabby-webserver/src/avatar.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/avatar.rs:23]: should add image mime type as response header.

File: ee/tabby-db/src/users.rs
Comment (boxbeam) [reply ee/tabby-db/src/users.rs:23]: Seems some of my changes got lost somewhere, I have just removed the field since it is never used as part of the user struct

File: ee/tabby-webserver/src/schema/mod.rs
Comment (boxbeam) [reply ee/tabby-webserver/src/schema/mod.rs:428]: Seems this was an issue of rebasing on main, though I don't know why the compiler error was showing here and not for me locally until I rebased

**Comment by wsxiaoys** [comment]
@wwayne please verify the api interface

File: ee/tabby-webserver/src/avatar.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/avatar.rs:28]:
```suggestion
        .insert(CONTENT_TYPE, "image/*".parse().unwrap());
```

File: ee/tabby-webserver/src/avatar.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/avatar.rs:13]: add unit test coverage

File: ee/tabby-webserver/src/schema/mod.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/mod.rs:429]: only user themselves can change their avatar, not even admin

File: ee/tabby-webserver/src/service/auth.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/service/auth.rs:1]: add unit test

**Review by wsxiaoys — APPROVED** [review]
Otherwise LGTM, pending on @wwayne confirmation on api

File: ee/tabby-webserver/graphql/schema.graphql
Comment (wwayne) [inline ee/tabby-webserver/graphql/schema.graphql:25]: good to me

**Review by wwayne — APPROVED** [review]

File: ee/tabby-webserver/src/avatar.rs
Comment (boxbeam) [reply ee/tabby-webserver/src/avatar.rs:13]: Please review commit `cadd57669020f429bfb0f568b4271500d84fcbef`, I had to change a number of things in order to get a test implemented

File: ee/tabby-webserver/src/service/auth.rs
Comment (boxbeam) [reply ee/tabby-webserver/src/service/auth.rs:1]: There already is a test within the auth service for setting an avatar

File: ee/tabby-webserver/src/avatar.rs
Comment (wsxiaoys) [reply ee/tabby-webserver/src/avatar.rs:13]: that's fine. will refine ast-grep rule a bit.

File: ee/tabby-webserver/src/schema/auth.rs
Comment (wsxiaoys) [inline ee/tabby-webserver/src/schema/auth.rs:466]:
```suggestion
    async fn update_user_avatar(&self, id: &ID, avatar: Option<Vec<u8>>) -> Result<()>;
```

Let's use blob format in service level

**Review by wsxiaoys — APPROVED** [review]
Otherwise LGTM

**Pull request merged**
